dev-support/design-docs/HBASE-23347-pluggable-authentication.md

```diff
@@ -0,0 +1,167 @@
+chapter for a tangent on SIMPLE authentication today)
+
+### Tangent: on SIMPLE authentication
+
+`SIMPLE` authentication in HBase today is treated as a special case. My impression is that
+this stems from HBase not originally shipping an RPC solution that had any authentication.
+
+Re-implementing `SIMPLE` authentication such that it also flows through SASL (e.g. via
+the `PLAIN` SASL mechanism) would simplify the HBase codebase such that all authentication
+occurs via SASL. This was not done for the initial implementation to reduce the scope
+of the changeset. Changing `SIMPLE` authentication to use SASL may result in some
+performance impact in setting up a new RPC. The same conditional logic to determine
+`if (sasl) ... else SIMPLE` logic is propagated in this implementation.
+
+## Implementation Overview
+
+HBASE-23347 includes a refactoring of HBase RPC authentication where all current methods
+are ported to a new set of interfaces, and all RPC implementations are updated to use
+the new interfaces. In the spirit of SASL, the expectation is that users can provide
+their own authentication methods at runtime, and HBase should be capable of negotiating
+a client who tries to authenticate via that custom authentication method. The implementation
+refers to this "bundle" of client and server logic as an "authentication provider".
+
+### Providers
+
+One authentication provider includes the following pieces:
+
+1. Client-side logic (providing a credential)
+2. Server-side logic (validating a credential from a client)
+3. Client selection logic to choose a provider (from many that may be available)
+
+A provider's client and server side logic are considered to be one-to-one. A `Foo` client-side provider
+should never be used to authenticate against a `Bar` server-side provider.
+
+We do expect that both clients and servers will have access to multiple providers. A server may
+be capable of authenticating via methods which a client is unaware of. A client may attempt to authenticate
+against a server which the server does not know how to process. In both cases, the RPC
+should fail when a client and server do not have matching providers. The server identifies
+client authentication mechanisms via a `byte authCode` (which is already sent today with HBase RPCs).
+
+A client may also have multiple providers available for it to use in authenticating against
+HBase. The client must have some logic to select which provider to use. Because we are
+allowing custom providers, we must also allow a custom selection logic such that the
+correct provider can be chosen. This is a formalization of the logic already present
+in `org.apache.hadoop.hbase.security.token.AuthenticationTokenSelector`.
+
+To enable the above, we have three new classes to support the user extensibility:
+
+1. `interface SaslClientAuthenticationProvider`
+2. `interface SaslServerAuthenticationProvider extends SaslClientAuthenticationProvider`
```

Comment:
Thinking through this, a concrete implementation for Kerberos would look something like..

* `abstract class GssSaslAuthenticationProvider implements SaslAuthenticationProvider`
* `class GssSaslClientAuthenticationProvider extends GssSaslAuthenticationProvider implements SaslClientAuthenticationProvider`
* `class GssSaslServerAuthenticationProvider extends GssSaslAuthenticationProvider implements SaslServerAuthenticationProvider`

We put the common logic in `GssSaslAuthenticationProvider` and just expect that other implementations will put their common logic in a base-class, rather than copy it into the client/server implementation.

My goal with the `server extends client` was to "force" a common hierarchy, but I guess I should also just trust that folks writing these will be smart :)